File: backend/app/schemas/memo.py
"""
公告系統 Pydantic Schemas
"""
from __future__ import annotations
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel


class MemoFileOut(BaseModel):
    id: str
    orig_name: str
    content_type: str
    size_bytes: int
    uploaded_by: str
    uploaded_at: datetime

    class Config:
        from_attributes = True


class MemoListItem(BaseModel):
    id: str
    title: str
    preview: str          # body 前 160 字（純文字）
    visibility: str
    author: str
    doc_no: str
    recipient: str
    source: str
    source_id: str
    created_at: datetime

    class Config:
        from_attributes = True


class MemoListResponse(BaseModel):
    items: List[MemoListItem]
    total: int
    page: int
    per_page: int


class MemoDetail(BaseModel):
    id: str
    title: str
    body: str
    visibility: str
    author: str
    author_id: str
    doc_no: str
    recipient: str
    source: str
    source_id: str
    created_at: datetime
    updated_at: datetime
    attachments: List[MemoFileOut] = []

    class Config:
        from_attributes = True


class MemoCreate(BaseModel):
    title: str
    body: str = ""
    visibility: str = "org"      # org | restricted
    doc_no: str = ""
    recipient: str = ""


class MemoUpdate(BaseModel):
    title: Optional[str] = None
    body: Optional[str] = None
    visibility: Optional[str] = None
    doc_no: Optional[str] = None
    recipient: Optional[str] = None

File: frontend/src/types/memo.ts
/**
 * 公告系統 TypeScript 型別定義
 * 對應後端 /api/v1/memos
 */

export type MemoVisibility = 'org' | 'restricted'

export interface MemoFileItem {
  id: string
  orig_name: string
  content_type: string
  size_bytes: number
  uploaded_by: string
  uploaded_at: string
}

export interface MemoListItem {
  id: string
  title: string
  preview: string         // 前 160 字純文字
  visibility: MemoVisibility
  author: string
  doc_no: string
  recipient: string
  source: string          // 'manual' | 'approval'
  source_id: string
  created_at: string
}

export interface MemoListResponse {
  items: MemoListItem[]
  total: number
  page: number
  per_page: number
}

export interface MemoDetail {
  id: string
  title: string
  body: string            // HTML / 純文字
  visibility: MemoVisibility
  author: string
  author_id: string
  doc_no: string
  recipient: string
  source: string
  source_id: string
  created_at: string
  updated_at: string
  attachments: MemoFileItem[]
}

export interface MemoCreatePayload {
  title: string
  body: string
  visibility: MemoVisibility
  doc_no: string
  recipient: string
}

export interface MemoUpdatePayload {
  title?: string
  body?: string
  visibility?: MemoVisibility
  doc_no?: string
  recipient?: string
}

export interface MemoFilters {
  q?: string
  visibility?: 'all' | MemoVisibility
  page?: number
  per_page?: number
}
